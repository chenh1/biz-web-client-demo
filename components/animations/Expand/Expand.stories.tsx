import type { ExpandType } from './Expand';
import React from 'react';
import { Expand } from './Expand';
import { Box } from '../../containers/Box/Box';
import { Heading } from '../../core/Heading/Heading';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { Button } from '../../core/Button/Button';

type Args = {
  args: ExpandType
}

const Template = (args) => (
  <Expand className={"max-w-xs h-[168px]"} {...args}>
    <Box className={"max-w-sm h-[168px]"} hasShadow isRounded gap="sm">
      <Heading type="h3">Heading text</Heading>
      <Paragraph>Example description</Paragraph>
      <Button label="CLICK ME" primary/>
    </Box>
  </Expand>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const ExpandAnimationHorizontal: Args = { args: { disable: false, orientation: "horizontal" } };
export const ExpandAnimationVertical: Args = { args: { disable: false, orientation: "vertical" } };


