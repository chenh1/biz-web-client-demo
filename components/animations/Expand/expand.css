@keyframes expandHorizontal {
  0% {
    width: 0;
  }
  60% {
    width: 105%;
  }
  80% {
    width: 98%;
  }
  100% {
    width: 100%;
  }
}

@keyframes expandVertical {
  0% {
    height: 0;
  }
  80% {
    height: 105%;
  }
  90% {
    height: 98%;
  }
  100% {
    height: 100%;
  }
}

.expand > * {
  overflow: hidden;
}

.expand > * *{
  white-space: nowrap;
}

.expand.horizontal > * {
  animation: expandHorizontal 1s ease-in-out;
}

.expand.vertical > * {
  animation: expandVertical 1s ease-in-out;
}

.freeze {
  animation-play-state: paused !important;
}