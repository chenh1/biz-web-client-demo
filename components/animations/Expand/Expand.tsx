'use client'

import type { Orientation, AnimationBase } from '../../constants/types';
import React, { useState } from 'react';
import { useIntersectionObserver } from '../../hooks/useIntersectionObserver';
import "./expand.css"

export type ExpandType = AnimationBase & {
  orientation: Orientation
}

export const Expand: React.FC<ExpandType> = ({ className = "", disable, animateWhenInView = true, orientation = "horizontal", children }) => {
  const [ isInView, setIsInView ] = useState(false);

  const moduleRef = useIntersectionObserver({
    inViewCallback: () => {
      setIsInView(true)
    },
    leaveViewCallback: () => {
      setIsInView(false)
    }
  })

  return (
    <div ref={moduleRef} className={`${!disable ? 'expand' : ''} ${!animateWhenInView || (animateWhenInView && isInView) ? '' : 'freeze'} ${orientation} ${className}`}>
      {children}
    </div>
  )
};