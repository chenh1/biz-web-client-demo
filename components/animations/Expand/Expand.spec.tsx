import { Expand } from "./Expand";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const observe = jest.fn();
const unobserve = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
global.IntersectionObserver = jest.fn(function (callback, options) {
  return {
    observe,
    unobserve,
    root: null,
    rootMargin: '',
    thresholds: [],
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  };
});


it('expands when in view', async () => {
  const { getByTestId, asFragment } = render(
    <Expand disable={false} orientation="horizontal">
      <p>Expander</p>
    </Expand>
  );

  await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
});