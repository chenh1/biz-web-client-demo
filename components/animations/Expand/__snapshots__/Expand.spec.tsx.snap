// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expands when in view 1`] = `
<DocumentFragment>
  <div
    class="expand freeze horizontal "
  >
    <p>
      Expander
    </p>
  </div>
</DocumentFragment>
`;
